accord-core/src/main/java/accord/local/CommandStores.java

```diff
@@ -148,12 +135,14 @@ public RangesForEpoch(long[] epochs, Ranges[] ranges, CommandStore store)
             this.store = store;
         }
 
-        public RangesForEpoch withRanges(long epoch, Ranges ranges)
+        public RangesForEpoch withRanges(long epoch, Ranges latestRanges)
         {
-            long[] newEpochs = Arrays.copyOf(this.epochs, this.epochs.length + 1);
-            Ranges[] newRanges = Arrays.copyOf(this.ranges, this.ranges.length + 1);
-            newEpochs[this.epochs.length] = epoch;
-            newRanges[this.ranges.length] = ranges;
+            Invariants.checkArgument(epochs.length == 0 || epochs[epochs.length - 1] <= epoch);
+            int newLength = epochs.length == 0 || epochs[epochs.length - 1] < epoch ? epochs.length + 1 : epochs.length;
+            long[] newEpochs = Arrays.copyOf(epochs, newLength);
+            Ranges[] newRanges = Arrays.copyOf(ranges, newLength);
+            newEpochs[newLength - 1] = epoch;
```

Comment:
In trying to understand this logic, it looks like its very dependent on the caller calling things in the correct order... would be good to add the following check

```suggestion
            Invariants.checkState(newEpochs[newLength - 1] == 0 || newEpochs[newLength - 1] == epoch, "Attempted to override historic epoch %d with %d", newEpochs[newLength - 1], epoch);
            newEpochs[newLength - 1] = epoch;
```